# 3. Data Manipulation I: Basics - Quiz

## 3.0. Import `polars` and Load Data

In [1]:
import polars as pl

In [2]:
df = pl.read_csv(
    "../data/yellow_tripdata_2024-03.csv",
    schema_overrides={"tpep_pickup_datetime": pl.Datetime, "tpep_dropoff_datetime": pl.Datetime}
)

## 3.1 Question 1: Toll Payment Column Datatype

Using `.select()`, fetch a column from `df` that represents whether or not a toll was paid as part of the trip. What is the datatype of that new column? (Hint: You can check whether a toll was paid by seeing if `tolls_amount` is greater than 0.)

In [3]:
result = (
    df
    .select(
        ### YOUR CODE HERE
    )
)
print(result)

shape: (3_582_628, 1)
┌──────────────┐
│ tolls_amount │
│ ---          │
│ bool         │
╞══════════════╡
│ false        │
│ false        │
│ false        │
│ false        │
│ false        │
│ …            │
│ false        │
│ false        │
│ false        │
│ false        │
│ false        │
└──────────────┘


1. `int64`
2. `float64`
3. `str`
4. `bool`

## 3.2 Question 2: Longest Trip with Zero Tolls

What was the longest trip that had zero tolls paid? (Hint: Use `.filter()` to get only the trips with `tolls_amount` equal to 0 and use `.select()` with `.max()` to find the longest trip.)

In [4]:
result = (
    df
    .filter(
        #### YOUR CODE HERE
    )
    .select(
        #### YOUR CODE HERE
    )
)
print(result)

shape: (1, 1)
┌───────────────┐
│ trip_distance │
│ ---           │
│ f64           │
╞═══════════════╡
│ 176836.3      │
└───────────────┘


1. 176836.3
2. 0
3. 176744.79
4. 176329.23

## 3.3 Question 3: Mean Tip for High Fare Trips

What is the mean tip amount for trips where the fare amount was greater than $20?

In [5]:
result = (
    df
    .filter(
        #### YOUR CODE HERE
    )
    .select(
        #### YOUR CODE HERE
    )
)
print(result)

shape: (1, 1)
┌─────────────┐
│ fare_amount │
│ ---         │
│ f64         │
╞═════════════╡
│ 39.32003    │
└─────────────┘


1. 39.32003
2. 598.58
3. 5.870243
4. 5.872741

## 3.4 Question 4: Max Trip Distance for 1-2 Passengers

Find the maximum distance for trips with a passenger count of 1 or 2.

In [6]:
result = (
    df
    .filter(
        #### YOUR CODE HERE
    )
    .select(
        #### YOUR CODE HERE
    )
)
print(result)

shape: (1, 1)
┌──────────────┐
│ total_amount │
│ ---          │
│ f64          │
╞══════════════╡
│ 1021.99      │
└──────────────┘


1. 159.74
2. 66907.9
3. 3.530788
4. 1021.99

## 3.5 Question 5: Top 5 Trips by Total Amount

Sort the DataFrame by `total_amount` in descending order. Then select and display only the top five rows and the columns `trip_distance` and `total_amount`. What are the two values of `trip_distance` associated with the two trips with the highest `total_amount`?

In [7]:
result = (
    df
    .sort(
        #### YOUR CODE HERE
    )
    .select([
        #### YOUR CODE HERE
    ])
    .head()  # Displays the top 5 rows
)
print(result)

shape: (5, 2)
┌───────────────┬──────────────┐
│ trip_distance ┆ total_amount │
│ ---           ┆ ---          │
│ f64           ┆ f64          │
╞═══════════════╪══════════════╡
│ 3.8           ┆ 1021.99      │
│ 181.5         ┆ 951.26       │
│ 0.0           ┆ 851.0        │
│ 0.1           ┆ 841.0        │
│ 159.74        ┆ 807.75       │
└───────────────┴──────────────┘


1. 0.0, 159.74
2. 3.8, 181.5
3. 1021.99, 951.26
4. 5.1, 8.3

## 3.8 Question 8: Latest Pickup for 5-Mile Trips

Of all the trips with a `trip_distance` of exactly 5, what is the latest `tpep_pickup_datetime`?

In [10]:
result = (
    df
    #### YOUR CODE HERE
)
print(result)

shape: (1, 1)
┌──────────────────────┐
│ tpep_pickup_datetime │
│ ---                  │
│ datetime[μs]         │
╞══════════════════════╡
│ 2024-03-31 23:46:21  │
└──────────────────────┘


1. 2024-03-31 23:46:21
2. 2024-03-31 23:57:07
3. 2024-03-01 00:05:00
4. 2024-04-01 00:34:55

## 3.9 Question 9: Lowest Tip Percentage

On those trips with a `fare_amount` and a `tip_amount` greater than 0, what is the lowest tip percentage (expressed as a fraction) that somebody paid? (Hint: Divide `tip_amount` by `fare_amount`.)

In [11]:
result = (
    df
    #### YOUR CODE HERE
)
print(result)

shape: (1, 1)
┌────────────────────┐
│ min_tip_percentage │
│ ---                │
│ f64                │
╞════════════════════╡
│ 0.00003            │
└────────────────────┘


1. 0.01
2. 0.276066
3. -40.0
4. 0.00003

## 3.10 Question 10: Trips with Tips Exceeding Airport Fee

How many trips have a `tip_amount` greater than the `Airport_fee`?

In [12]:
result = (
    df
    #### YOUR CODE HERE
)
print(result)

(2461463, 19)


1. 3582628
2. 2
3. 0
4. 2461463